src/transaction/transaction_manager.cpp

```diff
@@ -55,9 +56,44 @@ timestamp_t TransactionManager::ReadOnlyCommitCriticalSection(TransactionContext
   return commit_time;
 }
 
+bool TransactionManager::CheckConstraints(TransactionContext *txn,
+                                          std::vector<TransactionConstraint *> *txn_installed_constraints) {
+  auto iter = constraints_.Begin();
+  while (iter != constraints_.End()) {
+    TransactionConstraint &constraint = *iter;
+    if (constraint.InstallingTransactionId() == txn->TxnId().load()) {
+      txn_installed_constraints->push_back(&constraint);
+      constraint.SetEnforcing();
+      if (constraint.Violated()) {
+        return false;
+      }
+    } else if (!constraint.CheckConstraint(txn)) {
+      return false;
+    }
+    iter++;
+  }
+  return true;
+}
+
 timestamp_t TransactionManager::UpdatingCommitCriticalSection(TransactionContext *const txn, const callback_fn callback,
                                                               void *const callback_arg) {
   common::SharedLatch::ScopedExclusiveLatch guard(&commit_latch_);
+  std::vector<TransactionConstraint *> constraints_installed_by_txn;
+
+  if (!CheckConstraints(txn, &constraints_installed_by_txn)) {
+    // TODO(Yashwanth):
+    // Determine how to notify if transaction is aborted, current assumption seems to be if Commit is
+    // called then transaction can't be aborted
+
+    // Since constraint check failed, need to reset all the constraints belonging to this txn
+    for (TransactionConstraint *constraint : constraints_installed_by_txn) {
+      constraint->ResetEnforcing();
+    }
+
+    Abort(txn);
+    return timestamp_t(0);
```

Comment:
Probably returning the transaction's ID is a good choice because it is an invalid timestamp (highest order bit is 1).